lib/matplotlib/axes/_axes.py

```diff
@@ -5093,103 +5093,135 @@ def imshow(self, X, cmap=None, norm=None, aspect=None,
+            This parameter is ignored for RGBA input data.
 
-        origin : ['upper' | 'lower'], optional, default: None
+        origin : {'upper', 'lower'}, optional
             Place the [0,0] index of the array in the upper left or lower left
-            corner of the axes. If None, default to rc `image.origin`.
+            corner of the axes. The convention 'upper' is typically used for
+            matrices and images.
+            Defaults to :rc:`image.origin`.
 
-        extent : scalars (left, right, bottom, top), optional, default: None
+        extent : scalars (left, right, bottom, top), optional
             The location, in data-coordinates, of the lower-left and
-            upper-right corners. If `None`, the image is positioned such that
-            the pixel centers fall on zero-based (row, column) indices.
+            upper-right corners.
+
+            By default, pixels have a size of (1, 1). They are centered on
+            integer coordinates and their center coordinates range from 0 to
+            columns-1 horizontally and from 0 to rows-1 vertically.
+
+            Note that the meaning of *top* and *bottom* depend on *origin*
+            because *origin* changes the direction of the vertical axis.
+
+            - If origin is set to "upper", the center of the lower left
+              pixel is at (0, rows) , this corresponds to an extent
+              of (-0.5, columns-0.5, rows-0.5, -0.5).
+            - If origin is set to "lower", the center of the lower left
+              pixel is at (0, 0). This is equivalent to an extent
+              of (-0.5, columns-0.5, -0.5, rows-0.5).
 
         shape : scalars (columns, rows), optional, default: None
-            For raw buffer images
+            For raw buffer images.
+
+        filternorm : {0, 1}, optional, default: 1
+            Determines, whether the antigrain image resize filter will
+            normalize integer values.
 
-        filternorm : scalar, optional, default: 1
-            A parameter for the antigrain image resize filter.  From the
-            antigrain documentation, if `filternorm` = 1, the filter
+            From the antigrain documentation: if *filternorm* = 1, the filter
             normalizes integer values and corrects the rounding errors. It
             doesn't do anything with the source floating point values, it
             corrects only integers according to the rule of 1.0 which means
             that any sum of pixel weights must be equal to 1.0.  So, the
             filter function must produce a graph of the proper shape.
 
-        filterrad : scalar, optional, default: 4.0
+        filterrad : float > 0, optional, default: 4.0
             The filter radius for filters that have a radius parameter, i.e.
-            when interpolation is one of: 'sinc', 'lanczos' or 'blackman'
+            when interpolation is one of: 'sinc', 'lanczos' or 'blackman'.
+
+        resample
+            Undocumented.
```

Comment:
can we sort out what `resample` does?

Reply:
I couldn't. I can track down its use to _image_interpolation.h l.987. but that c code doesn't speak to me. It appears to be written by @mdboom. Maybe he could enlighten us?

Reply:
It is documented in [the resample C wrapper](https://github.com/matplotlib/matplotlib/blob/master/src/_image_wrapper.cpp#L38).

Reply:
Thanks. I've copied the doc. Still I'm not 100% clear what this does:
- When will resampling (no matter if full or partial) occur.
- what are the output and the input images?